README.md
# SMLogger

Powerful and customizable log manager for your Apps.


## Requirements

- iOS 9.0+ | macOS 10.10+ | tvOS 9.0+ | watchOS 2.0+


## Integration

#### Swift Package Manager

You can use [The Swift Package Manager](https://swift.org/package-manager) to install `SMLogger` by adding the proper description to your `Package.swift` file:

```swift
// swift-tools-version:5.5
import PackageDescription

let package = Package(
    name: "YOUR_PROJECT_NAME",
    dependencies: [
        .package(url: "https://github.com/siginur/SMLogger.git", from: "1.5.0"),
    ]
)
```
Then run `swift build` whenever you get prepared.


## Usage

### Initialization

```swift
import SMLogger
```
You can choose between predefined types of message format and output or to create your own by implementing `LogMessageFormat` and `LogOutput` protocols.
```swift
let logger = SMLogger(strategy: .default)
```
You can also to log by a few strategies at a time, just initialize `SMLogger` with an array of strategies.
```swift
let logger = SMLogger(strategies: [.default])
```

### LogStrategy
`LogStrategy` is a class that contains message format, output strategy for logging(e.x. you can log to console, file, text stream etc.), severity level filters and other preferences. <br>
By default `SMLogger` uses `.default` strategy that configured with `TextLogMessageFormat` and `ConsoleLogOutput`.

### LogSeverity
| Severity | Tag   | Level |
|:---------|-------|:------|
| fatal    | FATAL | 1000  |
| error    | ERROR | 800   |
| warning  | WARN  | 600   |
| info     | INFO  | 400   |
| debug    | DEBUG | 200   |
| trace    | TRACE | 100   |

Each log message has it's own severity level, you can filter logs by severity level in `LogStrategy` class.
Also you can create your own log severity levels.

### LogMessageFormat
`LogMessageFormat` protocol defines a format of your log messages, like date, line number, message etc. <br>
You have 2 predefined log message formats:
- `JSONLogMessageFormat` log messages in JSON format.
- `TextLogMessageFormat` log messages as a plain text. **[default]**

You can also create your own message format by implementing `LogMessageFormat` protocol.

### LogOutput
`LogOutput` protocol defines an output destination of your log messages.
There are 5 predefined log outputs:
- `ConsoleLogOutput` prints logs to console. **[default]**
- `FileLogOutput` save logs into file (can be configured to max number of lines per file or file size).
- `StreamLogOutput` streams to `TextOutputStream`.
- `NotificationLogOutput` send your logs via `NotificationCenter`.
- `MemoryLogOutput` store logs in memory.

You can also create your own log outputs by implementing `LogOutput ` protocol.













































CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/siginur/SMLogger/tree/HEAD)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.7.0...HEAD)

## [1.7.0](https://github.com/siginur/SMLogger/tree/v1.7.0) (2023-02-16)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.6.0...v1.7.0)
#### News:
- Add overridable `shouldPerform` method in `LogStrategy` class. 
- Add `extraInfo` parameter for all log methods.

## [1.6.0](https://github.com/siginur/SMLogger/tree/v1.6.0) (2022-04-22)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.5.0...v1.6.0)
#### News:
- Add possibility to create custom severity log levels.
- Add `MemoryLogOutput`.

## [1.5.0](https://github.com/siginur/SMLogger/tree/v1.5.0) (2022-04-09)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.4.0...v1.5.0)
#### News:
- Refactor and rename almost all classes.
- Add `active` flag to `LogStrategy` for disabling logs in runtime.
- Add readme and changelog files.
#### Changes:
- Change log severity levels.

## [1.4.0](https://github.com/siginur/SMLogger/tree/v1.4.0) (2021-12-15)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.3.0...v1.4.0)
#### News:
- Add column number to `LogMessageSegment`.
- Add methods for logging with string format.
- Add `NotificationLogOutputStrategy`.

## [1.3.0](https://github.com/siginur/SMLogger/tree/v1.3.0) (2021-12-14)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.2.4...v1.3.0)
#### News:
- Add `LogMessageFormat` and `LogMessageSegment`.

## [1.2.4](https://github.com/siginur/SMLogger/tree/v1.2.4) (2021-12-12)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.2.0...v1.2.4)
#### Fixes:
- Fix bugs in `FileLogOutputStrategy`.
- Fix multithreading bugs.

## [1.2.0](https://github.com/siginur/SMLogger/tree/v1.2.0) (2021-12-11)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.1.0...v1.2.0)
#### News:
- Replace old configurations by `LogMessageStrategy` and `LogOutputStrategy`.
- Add severity levels.

## [1.1.0](https://github.com/siginur/SMLogger/tree/v1.1.0) (2020-08-18)
[Full changelog](https://github.com/siginur/SMLogger/compare/v1.0.0...v1.1.0)
#### News:
- Add error logs.
- Add simple configurations.
#### Changes:
- Rename to `SMLogger`.

## [1.0.0](https://github.com/siginur/SMLogger/tree/v1.0.0) (2020-08-08)
Initial release.


Sources/SMLogger/LogSeverity.swift
//
//  LogLevel.swift
//  
//
//  Created by Alexey Siginur on 18/08/2020.
//

public struct LogSeverity: Comparable, Hashable {
    
    public static let fatal = LogSeverity(tag: "fatal", priority: 1000)
    public static let error = LogSeverity(tag: "error", priority: 800)
    public static let warning = LogSeverity(tag: "warn", priority: 600)
    public static let info = LogSeverity(tag: "info", priority: 400)
    public static let debug = LogSeverity(tag: "debug", priority: 200)
    public static let trace = LogSeverity(tag: "trace", priority: 100)

    let tag: String
    let priority: Int
    
    public static func >=(lhs: LogSeverity, rhs: LogSeverity) -> Bool {
        return lhs.priority >= rhs.priority
    }

    public static func <(lhs: LogSeverity, rhs: LogSeverity) -> Bool {
        return lhs.priority < rhs.priority
    }
    
}

public enum LogSeverityFilter {
    case all
    case min(LogSeverity)
    case max(LogSeverity)
    case range(ClosedRange<LogSeverity>)
    case exclude(Set<LogSeverity>)
    case include(Set<LogSeverity>)
    
    public func pass(_ logSeverity: LogSeverity) -> Bool {
        switch self {
        case .all:
            return true
        case .min(let min):
            return logSeverity >= min
        case .max(let max):
            return logSeverity <= max
        case .range(let range):
            return range.contains(logSeverity)
        case .exclude(let exclude):
        	return exclude.contains(logSeverity) == false
        case .include(let include):
        	return include.contains(logSeverity) == true
        }
    }
}
